components/brave_wallet/browser/zcash/zcash_orchard_storage.cc

```diff
@@ -0,0 +1,438 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_orchard_storage.h"
+
+#include "base/files/file_util.h"
+#include "base/strings/stringprintf.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "sql/statement.h"
+#include "sql/transaction.h"
+
+namespace brave_wallet {
+
+namespace {
+#define kNotesTable "notes"
+#define kSpentNotesTable "spent_notes"
+#define kAccountMeta "account_meta"
+
+const int kEmptyDbVersionNumber = 1;
+const int kCurrentVersionNumber = 2;
+
+std::optional<uint32_t> ReadUint32(sql::Statement& statement, size_t index) {
+  auto v = statement.ColumnInt64(index);
+  if (v < 0 || v > std::numeric_limits<uint32_t>::max()) {
+    return std::nullopt;
+  }
+  return static_cast<uint32_t>(v);
+}
+
+}  // namespace
+
+ZCashOrchardStorage::ZCashOrchardStorage(base::FilePath path_to_database)
+    : db_file_path_(path_to_database), database_() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+}
+
+ZCashOrchardStorage::~ZCashOrchardStorage() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+}
+
+bool ZCashOrchardStorage::EnsureDbInit() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  if (database_.is_open()) {
+    return true;
+  }
+  return CreateOrUpdateDatabase();
+}
+
+void ZCashOrchardStorage::ResetDatabase() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  database_.Close();
+  sql::Database::Delete(db_file_path_);
+}
+
+bool ZCashOrchardStorage::CreateOrUpdateDatabase() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  const base::FilePath dir = db_file_path_.DirName();
+  if (!base::DirectoryExists(dir) && !base::CreateDirectory(dir)) {
+    NOTREACHED_IN_MIGRATION();
+    return false;
+  }
+
+  if (!database_.Open(db_file_path_)) {
+    NOTREACHED_IN_MIGRATION();
+    return false;
+  }
+
+  sql::MetaTable meta_table;
+  if (!meta_table.Init(&database_, kEmptyDbVersionNumber,
+                       kEmptyDbVersionNumber)) {
+    NOTREACHED_IN_MIGRATION();
+    database_.Close();
+    return false;
+  }
+
+  if (meta_table.GetVersionNumber() == kEmptyDbVersionNumber) {
+    if (!CreateSchema() ||
+        !meta_table.SetVersionNumber(kCurrentVersionNumber)) {
+      NOTREACHED_IN_MIGRATION();
+      database_.Close();
+      return false;
+    }
+  } else if (meta_table.GetVersionNumber() < kCurrentVersionNumber) {
+    if (!UpdateSchema() ||
+        !meta_table.SetVersionNumber(kCurrentVersionNumber)) {
+      NOTREACHED_IN_MIGRATION();
+      database_.Close();
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool ZCashOrchardStorage::CreateSchema() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  sql::Transaction transaction(&database_);
+  return transaction.Begin() &&
```

Comment:
Would it make sense for us to encrypt this db/tables with the wallet password? AFAICT, librustzcash implementation isn't doing this and it doesn't seem like knowledge of the notes would compromise the protocol if the attacker also wasn't able to gain access to the keys, but I'm thinking it might be useful as a defense in depth measure. What's your thoughts on this and how complex would it be to do?

Reply:
I don't see any native encryption support in //sql. This would need more additional research.

Reply:
https://github.com/brave/brave-browser/issues/40256

Reply:
Fine by me to address as follow up - This also I don't think presents any security concerns but is a slight privacy concern. Normally we wouldn't consider reading from the disk apart of the threat model, but this seems like a good defense in depth measure.